src/development/perf-benchmarking.md

```diff
@@ -0,0 +1,117 @@
+# Library optimizations and benchmarking
+
+Recommended reading: [The Rust performance book](https://nnethercote.github.io/perf-book/title-page.html)
+
+## What to optimize
+
+It's preferred to optimize code that shows up as significant in real-world code.
+E.g. it's more beneficial to speed up `[T]::sort` than it is to shave off a small allocation in `Command::spawn`
+because the latter is dominated by its syscall cost.
+
+Issues about slow library code are labeled as [I-slow T-libs](https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+label%3AI-slow+label%3AT-libs)
+and those about code size as [I-heavy T-libs](https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+label%3AI-heavy+label%3AT-libs)
+
+## Vectorization
+
+Currently explicit SIMD features can't be used in alloc or core because runtime feature-detection is only available in std
+and they are compiled with each target's baseline feature set.
+
+Vectorization can only be achieved by shaping code in a way that the compiler backend's auto-vectorization passes can understand.
+
+## rustc-perf
+
+For parts of the standard library that are heavily used by rustc itself it can be convenient to use
+[the benchmark server](https://github.com/rust-lang/rustc-perf/tree/master/collector#benchmarking).
+
+Since it only measures compile-time but not runtime performance of crates it can't be used to benchmark for features
+that aren't used by the compiler, e.g. floating point code, linked lists, mpsc channels, etc.
+For those explicit benchmarks must be written or extracted from real-world code.
+
+## Built-in Microbenchmarks
+
+The built-in benchmarks use [cargo bench](https://doc.rust-lang.org/nightly/unstable-book/library-features/test.html)
+and can be found in the `benches` directory for `core` and `alloc` and in `test` modules in `std`.
+
+The benchmarks are automatically executed run in a loop by `Bencher::iter` to average the runtime over many loop-iterations.
+For CPU-bound microbenchmarks the runtime of a single iteration should be in the range of nano- to microseconds.
+
+To run a specific  can be invoked without recompiling rustc
+via `./x bench library/<lib> --stage 0 --test-args <benchmark name>`.
+
+`cargo bench` measures wall-time. This often is good enough, but small changes such as saving a few instructions
+in a bigger function can get drowned out by system noise. In such cases the following changes can make runs more
+reproducible:
+
+* disable incremental builds in `config.toml`
+* build std and the benchmarks with `RUSTFLAGS_BOOTSTRAP="-Ccodegen-units=1"`
+* ensure the system is as idle as possible
+* [disable ASLR](https://man7.org/linux/man-pages/man8/setarch.8.html)
+* [pinning](https://man7.org/linux/man-pages/man1/taskset.1.html) the benchmark process to a specific core
+* [disable clock boosts](https://wiki.archlinux.org/title/CPU_frequency_scaling#Configuring_frequency_boosting),
+  especially on thermal-limited systems such as laptops
```

Comment:
We can also add a link to something like https://github.com/JuliaCI/BenchmarkTools.jl/blob/master/docs/src/linuxtips.md.

Reply:
Some of those things may not be relevant to std benchmarks, which are mostly are CPU- or memory-bandwidth-bound. They shouldn't suffer much from swap, IRQs or SMT-siblibgs if you ensured the system is mostly idle since they depend on system activity (well, depends on how many cores one has... maybe core isolation is still worth it). 

Scheduling and throttling have the biggest impact in my experience. If we had a benchmark that tried to do a parallel sort on a huge dataset that would be a different story.

Adjusting the scaling governor is a good point.